src/sentry/utils/eventuser.py

```diff
@@ -96,59 +104,160 @@ def for_projects(
                 )
+            )
 
-            else:
-                where_conditions.append(Condition(Column(snuba_column), Op.EQ, value))
+        if len(keyword_where_conditions) == 1:
+            where_conditions.extend(
+                keyword_where_conditions,
+            )
+
+        columns = [
+            Column("project_id"),
+            Column("group_id"),
+            Column("ip_address_v6"),
+            Column("ip_address_v4"),
+            Column("user_id"),
+            Column("user"),
+            Column("user_name"),
+            Column("user_email"),
+        ]
 
         query = Query(
             match=Entity(EntityKey.Events.value),
             select=[
-                Column("project_id"),
-                Column("group_id"),
-                Column("ip_address_v6"),
-                Column("ip_address_v4"),
-                Column("event_id"),
-                Column("user_id"),
-                Column("user"),
-                Column("user_name"),
-                Column("user_email"),
+                *columns,
+                Function("max", [Column("timestamp")], "latest_timestamp"),
             ],
             where=where_conditions,
-            limit=Limit(1),
-            orderby=[OrderBy(Column("timestamp"), Direction.DESC)],
+            groupby=[*columns],
+            orderby=[OrderBy(Column("latest_timestamp"), Direction.DESC)],
         )
 
+        if not return_all:
+            query.set_limit(1)
+
         request = Request(
             dataset=Dataset.Events.value,
             app_id=REFERRER,
             query=query,
             tenant_ids={"referrer": REFERRER, "organization_id": projects[0].organization.id},
         )
         data_results = raw_snql_query(request, referrer=REFERRER)["data"]
-        results = [EventUser.from_snuba(result) for result in data_results]
+
+        # Return the first matching item from the Snuba results.
+        # All other rows are not needed.
+        # The Snuba results are sorted by descending time.
+        first_matching_items = []
+
+        snuba_keyword_filters = {}
+        for keyword, value in keyword_filters.items():
+            snuba_column = SNUBA_KEYWORD_MAP.get_key(keyword)
+            snuba_keyword_filters[snuba_column] = value
+
+        matches = self._find_matching_items(data_results, snuba_keyword_filters, filter_boolean)
+        first_matching_items.extend(matches.values())
+
+        results = [EventUser.from_snuba(item) for item in first_matching_items]
+
+        analytics.record(
+            "eventuser_snuba.query",
+            project_ids=[p.id for p in projects],
+            query=query.print(),
+            count_rows_returned=len(data_results),
+            count_rows_filtered=len(data_results) - len(results),
+        )
 
         return results
 
+    @staticmethod
+    def _find_matching_items(
+        snuba_results: List[dict[str, Any]], filters: Mapping[str, Any], filter_boolean: str
+    ):
```

Comment:
We don't need to match on the keyword filters. The way to find a unique `EventUser` is to look at the `sentry:user` tag, which maps to the `user` column in Snuba. The value is assigned here: https://github.com/getsentry/sentry/blob/27e72509103590238fca6a8bb88b9a1b7ac8ef7c/src/sentry/event_manager.py#L1064

So the `user` column will just be the result of https://github.com/getsentry/sentry/blob/fda93e57f0f2d311cfc7ecfda194236cf049a5f0/src/sentry/models/eventuser.py#L85-L91

That tag is mapped to a column here: https://github.com/getsentry/snuba/blob/94c47010b1c6425942684b54d3ead222c0ad181a/snuba/datasets/configuration/events/storages/errors.yaml#L350

So if we had a window function, we'd partition on `user` and fetch the first row for each user. In this case, we have to group on all the cols and just select the most recent user in code instead.